// @nextcloud/vue/dist/Functions/reference.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Functions_registerReference = require("../chunks/index-BV85rPB7.cjs");
const referencePickerModal = require("../chunks/referencePickerModal-qWVMGb2m.cjs");
exports.NcCustomPickerRenderResult = Functions_registerReference.NcCustomPickerRenderResult;
exports.hasInteractiveView = Functions_registerReference.hasInteractiveView;
exports.isCustomPickerElementRegistered = Functions_registerReference.isCustomPickerElementRegistered;
exports.isWidgetRegistered = Functions_registerReference.isWidgetRegistered;
exports.registerCustomPickerElement = Functions_registerReference.registerCustomPickerElement;
exports.registerWidget = Functions_registerReference.registerWidget;
exports.renderCustomPickerElement = Functions_registerReference.renderCustomPickerElement;
exports.renderWidget = Functions_registerReference.renderWidget;
exports.anyLinkProviderId = referencePickerModal.anyLinkProviderId;
exports.getLinkWithPicker = referencePickerModal.getLinkWithPicker;
exports.getProvider = referencePickerModal.getProvider;
exports.getProviders = referencePickerModal.getProviders;
exports.searchProvider = referencePickerModal.searchProvider;
exports.sortProviders = referencePickerModal.sortProviders;
