// @nextcloud/vue/dist/Functions/registerReference.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Functions_registerReference = require("../chunks/index-BV85rPB7.cjs");
exports.NcCustomPickerRenderResult = Functions_registerReference.NcCustomPickerRenderResult;
exports.destroyCustomPickerElement = Functions_registerReference.destroyCustomPickerElement;
exports.destroyWidget = Functions_registerReference.destroyWidget;
exports.getCustomPickerElementSize = Functions_registerReference.getCustomPickerElementSize;
exports.isCustomPickerElementRegistered = Functions_registerReference.isCustomPickerElementRegistered;
exports.isWidgetRegistered = Functions_registerReference.isWidgetRegistered;
exports.registerCustomPickerElement = Functions_registerReference.registerCustomPickerElement;
exports.registerWidget = Functions_registerReference.registerWidget;
exports.renderCustomPickerElement = Functions_registerReference.renderCustomPickerElement;
exports.renderWidget = Functions_registerReference.renderWidget;
