// @nextcloud/vue/dist/Components/NcRichText.cjs
"use strict";
Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: "Module" } });
const NcRichText = require("../chunks/NcRichText-BzJv-N1n.cjs");
const Functions_registerReference = require("../chunks/index-BV85rPB7.cjs");
const referencePickerModal = require("../chunks/referencePickerModal-qWVMGb2m.cjs");
exports.NcReferenceList = NcRichText.NcReferenceList;
exports.NcRichText = NcRichText.NcRichText;
exports.default = NcRichText.NcRichText;
exports.NcCustomPickerRenderResult = Functions_registerReference.NcCustomPickerRenderResult;
exports.isCustomPickerElementRegistered = Functions_registerReference.isCustomPickerElementRegistered;
exports.isWidgetRegistered = Functions_registerReference.isWidgetRegistered;
exports.registerCustomPickerElement = Functions_registerReference.registerCustomPickerElement;
exports.registerWidget = Functions_registerReference.registerWidget;
exports.renderCustomPickerElement = Functions_registerReference.renderCustomPickerElement;
exports.renderWidget = Functions_registerReference.renderWidget;
exports.NcReferencePicker = referencePickerModal.NcReferencePicker;
exports.NcReferencePickerModal = referencePickerModal.NcReferencePickerModal;
exports.NcReferenceWidget = referencePickerModal.NcReferenceWidget;
exports.NcSearch = referencePickerModal.NcSearch;
exports.anyLinkProviderId = referencePickerModal.anyLinkProviderId;
exports.getLinkWithPicker = referencePickerModal.getLinkWithPicker;
exports.getProvider = referencePickerModal.getProvider;
exports.getProviders = referencePickerModal.getProviders;
exports.searchProvider = referencePickerModal.searchProvider;
exports.sortProviders = referencePickerModal.sortProviders;
